docs/architecture/adr-038-state-listening.md

```diff
@@ -0,0 +1,584 @@
+# ADR 038: KVStore state listening
+
+## Changelog
+
+- 11/23/2020: Initial draft
+
+## Status
+
+Proposed
+
+## Abstract
+
+This ADR defines a set of changes to enable listening to state changes of individual KVStores and exposing these data to consumers.
+
+## Context
+
+Currently, KVStore data can be remotely accessed through [Queries](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/messages-and-queries.md#queries) which proceed through Tendermint and the ABCI.
+In addition to these request/response queries, it would be beneficial to have a means of listening to state changes as they occur in real time.
+
+## Decision
+
+We will modify the `MultiStore` interface and its concrete (`rootmulti` and `cachemulti`) implementations and introduce a new `listenkv.Store` to allow listening to state changes in underlying KVStores.
+We will also introduce the tooling for writing these state changes out to a file.
+
+### Listening interface
+
+In a new file- `store/types/listening.go`- we will create a `WriteListener` interface for streaming out state changes from a KVStore.
+
+```go
+// WriteListener interface for streaming data out from a listenkv.Store
+type WriteListener interface {
+  // if value is nil then it was deleted
+  OnWrite(key []byte, value []byte)
+}
+```
+
+### Listener type
+
+We will create two concrete implementation of the `WriteListener` interface in `store/types/listening.go`.
+
+One that writes out length-prefixed key-value pairs to an underlying `io.Writer`:
+
+```go
+// PrefixWriteListener is used to configure listening to a KVStore by writing out big endian length-prefixed
+// key-value pairs to an io.Writer
+type PrefixWriteListener struct {
+	writer io.Writer
+	prefixBuf [6]byte
+}
+
+// NewPrefixWriteListener wraps a PrefixWriteListener around an io.Writer
+func NewPrefixWriteListener(w io.Writer) *PrefixWriteListener {
+	return &PrefixWriteListener{
+		writer: w,
+	}
+}
+
+// OnWrite satisfies the WriteListener interface by writing out big endian length-prefixed key-value pairs
+// to an underlying io.Writer
+// The first two bytes of the prefix encode the length of the key
+// The last four bytes of the prefix encode the length of the value
+// This WriteListener makes two assumptions
+// 1) The key is no longer than 1<<16 - 1
+// 2) The value is no longer than 1<<32 - 1
```

Comment:
why not just write:
```
[key-length-prefix][key-bytes][value-length-prefix][value-bytes]{repeat}
```

and we can remove the 2 assumptions here.

Reply:
I think we'd still need to define a static length for the key-length and value-length prefixes, and then assume that all keys and values fit within those lengths. Or we could use additional 1 byte prefix-length prefixes for the key and value prefixes to inform their length.

e.g.

`[key-length-prefix-length-prefix][key-length-prefix][key][value-length-prefix-length-prefix][value-length-prefix][value]`

Reply:
I don't see why we'd need to do that. length-prefixed encoding should be fixed size, not variable. (e.g. int16)

Reply:
You could also have this just be a protobuf-based message.

Reply:
To clarify further, it's not the key bytes or value bytes themselves that are overflowing here, its their length prefixes that can potentially overflow. If the key is longer than `1<<16 - 1` we can't encode its length in the two byte prefix, if the value is longer than `1<<32-1` we can't encode its length in the the four byte prefix.

Reply:
During our regen call yesterday, varints were proposed to be used here as length prefixes. TBH this becomes similar to what protobuf does, so using protobuf also seems like a good idea to me (final binary size might be bigger, but we're already using protobuf everywhere so...)

Reply:
Thank you @alexanderbez and @amaurymartiny, protobuf it is :)